src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/IncrementalGenerationTests.cs

```diff
@@ -210,5 +209,52 @@ public async Task ChangingMarshallingAttributes_SameStrategy_DoesNotRegenerate()
                         output => Assert.Equal(IncrementalStepRunReason.Unchanged, output.Reason));
                 });
         }
+
+        [Fact]
+        public async Task GeneratorRun_WithNewCompilation_DoesNotKeepOldCompilationAlive()
+        {
+            string source = $"namespace NS{{{CodeSnippets.BasicParametersAndModifiers<int>()}}}";
+
+            SyntaxTree syntaxTree = CSharpSyntaxTree.ParseText(source, new CSharpParseOptions(LanguageVersion.Preview));
+
+            Compilation comp1 = await TestUtils.CreateCompilation(new[] { syntaxTree });
+
+            var (reference, driver) = RunTwoGeneratorOnTwoIterativeCompilationsAndReturnFirst(comp1);
+
+            GC.Collect();
+
+            Assert.False(reference.IsAlive);
+            GC.KeepAlive(driver);
+
+            static (WeakReference reference, GeneratorDriver driver) RunTwoGeneratorOnTwoIterativeCompilationsAndReturnFirst(Compilation startingCompilation)
```

Comment:
Just for sanity, perhaps we should mark this as "no inline". Given the size this is unlikely, but an implied invariant here is that this not be inlined.